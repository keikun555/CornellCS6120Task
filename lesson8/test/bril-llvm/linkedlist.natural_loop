{
  "cons": {
    "reducible": true,
    "backedges": []
  },
  "print_list": {
    "reducible": true,
    "backedges": []
  },
  "free_list": {
    "reducible": true,
    "backedges": []
  },
  "main": {
    "reducible": true,
    "backedges": []
  }
}
