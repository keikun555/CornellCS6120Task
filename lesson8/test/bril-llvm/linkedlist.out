@cons(head.0: int, tail.0: ptr<int_list>): ptr<int_list> {
.cons.b0:
  head.1: int = id head.0;
  tail.1: ptr<int_list> = id tail.0;
.cons.b1:
  one.0: int = const 1;
  p.0: ptr<int_list> = alloc one.0;
  phead.0: ptr<int> = getmbr p.0 elt;
  ptail.0: ptr<ptr<int_list>> = getmbr p.0 next;
  store phead.0 head.1;
  store ptail.0 tail.1;
  ret p.0;
}
@print_list(list.0: ptr<int_list>) {
.print_list.b0:
  list.1: ptr<int_list> = id list.0;
.print_list.b1:
  empty.0: bool = isnull list.1;
  br empty.0 .end .print;
.print:
  xp.0: ptr<int> = getmbr list.1 elt;
  x.0: int = load xp.0;
  print x.0;
  tp.0: ptr<ptr<int_list>> = getmbr list.1 next;
  t.0: ptr<int_list> = load tp.0;
  call @print_list t.0;
.end:
  xp.1: ptr<int> = phi xp.0 .print;
  x.1: int = phi x.0 .print;
  tp.1: ptr<ptr<int_list>> = phi tp.0 .print;
  t.1: ptr<int_list> = phi t.0 .print;
  ret;
}
@free_list(list.0: ptr<int_list>) {
.free_list.b0:
  list.1: ptr<int_list> = id list.0;
.free_list.b1:
  empty.0: bool = isnull list.1;
  br empty.0 .end .freetail;
.freetail:
  tp.0: ptr<ptr<int_list>> = getmbr list.1 next;
  t.0: ptr<int_list> = load tp.0;
  call @free_list t.0;
  free list.1;
.end:
  tp.1: ptr<ptr<int_list>> = phi tp.0 .freetail;
  t.1: ptr<int_list> = phi t.0 .freetail;
  ret;
}
@main {
.main.b0:
  a.0: int = const 2;
  b.0: int = const 3;
  c.0: int = const 5;
  d.0: int = const 8;
  n.0: ptr<int_list> = const 0;
  s0.0: ptr<int_list> = call @cons a.0 n.0;
  s1.0: ptr<int_list> = call @cons b.0 s0.0;
  s2.0: ptr<int_list> = call @cons c.0 s1.0;
  s3.0: ptr<int_list> = call @cons d.0 s2.0;
  call @print_list s3.0;
  call @free_list s3.0;
  ret;
}
